### PR TITLE
Clarify the "version" requirement in CRI and fix various bugs in dockershim

### LINKED ISSUE #32741: CRI: kuberuntime can't handle docker api version
umbrella issue: #31459

kuberuntime uses `semver` to ]parse the version string of the api version](https://github.com/kubernetes/kubernetes/blob/6538912013c16bf1e9c9b0ad1e418e9696849906/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L213) from the runtime. 
When using with docker, it complains that `1.23 is not in dotted-tri format`

Maybe we should just enforce that the api version returned by the runtime should be semver compatible.

/cc @feiskyer @Random-Liu @kubernetes/sig-node

### PR DESCRIPTION
**Pull request by yujuhong**
This fixes #32741

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32744)

### CODE REVIEW

File: pkg/kubelet/dockershim/docker_container.go [hunk 1]
```diff
@@ -103,7 +103,6 @@ func (ds *dockerService) CreateContainer(podSandboxID string, config *runtimeApi
 		Name: makeContainerName(sandboxConfig, config),
 		Config: &dockercontainer.Config{
 			// TODO: set User.
-			Hostname:   sandboxConfig.GetHostname(),
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/docker_container.go]: Why removing this?

Comment (yujuhong) [inline pkg/kubelet/dockershim/docker_container.go]: Because it's wrong. We only set the hostname for the infra container (aka the sandbox)

Comment (Random-Liu) [inline pkg/kubelet/dockershim/docker_container.go]: ACK

**Comment by Random-Liu** [comment]
LGTM

**Comment by Random-Liu** [comment]
BTW, there is new Project feature that we can leverage. :)

**Comment by yujuhong** [comment]
> BTW, there is a new Project feature that we can leverage. :)

Yes, we should definitely leverage that when we can.

**Comment by Random-Liu** [comment]
```
E0914 17:11:19.685984   20604 main.go:84] Error: Failed making a context: /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/dockershim/docker_container.go:35:11: undeclared name: dockerService
FAILED   hack/make-rules/../../hack/verify-import-boss.sh	19s
```

**Comment by yujuhong** [comment]
Fixed the check. Reapplying lgtm.

**Pull request merged**